Add LLaVa-1.6, bis

NielsRogge opened the pull request:
# What does this PR do?

This PR adds [LLaVa-1.6](https://llava-vl.github.io/blog/2024-01-30-llava-next/) (also called LLaVa-NeXT) as a standalone model rather than modifying the existing modeling_llava.py. As discussed in #29012 this might be better.

To do:

- [x] convert all checkpoints once PR is approved, along with corresponding integration test, which will also fix the failing documentation test.

NielsRogge committed:
First draft

NielsRogge committed:
Fix tests, add docs

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_29586). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

NielsRogge committed:
Improve docstrings

amyeroberts commented:
Thanks for opening @NielsRogge! I can see some model related tests are failing - ping me for review once they're all passing

NielsRogge committed:
Fix test

NielsRogge commented:
@amyeroberts I've fixed all tests, only the documentation test is failing because the official checkpoint hasn't been uploaded yet.

amyeroberts commented on src/transformers/models/llava_next/modeling_llava_next.py:
```diff
@@ -0,0 +1,684 @@
+                )
+                if labels is None:
+                    labels = torch.full_like(attention_mask, self.config.ignore_index).to(torch.long)
+            else:
+                # In case input_ids.shape[1] == 1 & pixel_values==None & past_key_values != None, we are in the case of
+                # generation with cache
+                if past_key_values is not None and pixel_values is not None and input_ids.shape[1] == 1:
```
What if any of these conditions aren't true - what situation are we in? If it's not valid, we should raise an error. For example, can I call this method directly with `input_embeds` if `input_ids` is None?

```suggestion
            elif past_key_values is not None and pixel_values is not None and input_ids.shape[1] == 1:
                # In case input_ids.shape[1] == 1 & pixel_values==None & past_key_values != None, we are in the case of
                # generation with cache
```

amyeroberts commented on src/transformers/models/llava_next/image_processing_llava_next.py:
```diff
@@ -0,0 +1,709 @@
+        list: A list of np.array representing the patches.
+    """
+    patches = []
+    height, width = image.shape[:-1] if input_data_format == ChannelDimension.LAST else image.shape[1:]
```
```suggestion
    height, width = get_image_size(image, input_data_format)
```

amyeroberts commented on src/transformers/models/llava_next/image_processing_llava_next.py:
```diff
@@ -0,0 +1,709 @@
+
+    Args:
+        aspect_ratio_setting (`str`, *optional*, defaults to `"anyres"`):
+            The aspect ratio setting to use. Can be "clip" (as in CLIP), "pad" (as in LLaVa 1.5) or "anyres" (as in LLaVa 1.6).
```
* I'm not sure it make sense to support the "pad" option now we have a new model 
* Under what cases should "clip" be used?

amyeroberts replied:
To be addressed

NielsRogge replied:
"clip" isn't really used (that's the default way of how CLIP preprocesses images), only "pad" is used (for llava 1.5) and "anyres" (for llava 1.6).

[People have reported](https://huggingface.co/llava-hf/llava-1.5-7b-hf/discussions/18) that llava 1.5-hf has the wrong preprocessing settings ("clip"), so they could use the "pad" setting if they want one-on-one matching results:

NielsRogge replied:
Removed this flag

amyeroberts commented on src/transformers/models/llava_next/image_processing_llava_next.py:
```diff
@@ -0,0 +1,709 @@
+            A list of possible resolutions to use for processing high resolution images. The best resolution is selected
+            based on the original size of the image. Can be overridden by `image_grid_pinpoints` in the `preprocess`
+            method.
+        resample (`PILImageResampling`, *optional*, defaults to `Resampling.BICUBIC`):
```
Double checking this matches the original implementation

NielsRogge replied:
Yes, original implementation also uses BICUBIC interpolation, which is asserted in the conversion script [here](https://github.com/NielsRogge/transformers/blob/add_llava_1_6_bis/src/transformers/models/llava_next/convert_llava_next_weights_to_hf.py#L174)

amyeroberts commented on src/transformers/models/llava_next/image_processing_llava_next.py:
```diff
@@ -0,0 +1,709 @@
+        original_height, original_width = get_image_size(image, channel_dim=input_data_format)
+        target_height, target_width = target_resolution
+
+        scale_w = target_width / original_width
+        scale_h = target_height / original_height
+
+        if scale_w < scale_h:
+            new_width = target_width
+            new_height = min(math.ceil(original_height * scale_w), target_height)
+        else:
+            new_height = target_height
+            new_width = min(math.ceil(original_width * scale_h), target_width)
```
As this logic is exactly the same as in `_resize_for_patching` we can abstract it out to a utility function something like `_get_patch_output_size(original_size, target_resolution)`

amyeroberts commented on src/transformers/models/llava_next/image_processing_llava_next.py:
```diff
@@ -0,0 +1,709 @@
+
+        patches = divide_to_patches(padded_image, patch_size=patch_size, input_data_format=input_data_format)
+
+        # make sure that all patches use the input data format
```
This comment isn't clear - the logic below is making sure all patches are in the `data_format` format, not 'using' the input data format

amyeroberts commented on src/transformers/models/llava_next/image_processing_llava_next.py:
```diff
@@ -0,0 +1,709 @@
+    def preprocess(
+        self,
+        images: ImageInput,
+        aspect_ratio_setting=None,
```
```suggestion
        aspect_ratio_setting: Optional[str] = None,
```

amyeroberts commented on tests/models/llava_next/test_modeling_llava_next.py:
```diff
@@ -0,0 +1,405 @@
+        },
+        is_training=True,
+        vision_config={
+            "batch_size": 12,
```
Why do we need batch_size here?

amyeroberts commented on tests/models/llava_next/test_modeling_llava_next.py:
```diff
@@ -0,0 +1,405 @@
+    def test_feed_forward_chunking(self):
+        pass
+
+    @unittest.skip(reason="CPU offload is not yet supported")
+    def test_cpu_offload(self):
+        pass
```
This is funny - if we have `_no_split_modules` set then it should be supported. Otherwise the model will automatically be skipped for this test, so we shouldn't need it

NielsRogge replied:
Pinging @younesbelkada here regarding how `_no_split_modules` should be defined for llava/llava-next.

amyeroberts commented on tests/models/llava_next/test_modeling_llava_next.py:
```diff
@@ -0,0 +1,405 @@
+    """
+
+    all_model_classes = (LlavaNextForConditionalGeneration,) if is_torch_available() else ()
+    fx_compatible = False
```
Was this tested?

NielsRogge replied:
This actually works, so removing this for both llava-next and llava

amyeroberts commented on tests/models/llava_next/test_modeling_llava_next.py:
```diff
@@ -0,0 +1,405 @@
+
+    @slow
+    @require_bitsandbytes
+    def test_small_model_integration_test_llama_batched(self):
```
mistral, based on the checkpoint 

```suggestion
    def test_small_model_integration_test_mistral_batched(self):
```

amyeroberts reviewed (commented):
Thanks for adding this as a new architecture - it's a lot cleaner this way!

NielsRogge committed:
Address comments

NielsRogge committed:
Address comments

NielsRogge commented on src/transformers/models/llava_next/processing_llava_next.py:
```diff
@@ -0,0 +1,136 @@
+            image_inputs = self.image_processor(images, return_tensors=return_tensors)
+        else:
+            image_inputs = {}
+        text_inputs = self.tokenizer(
+            text, return_tensors=return_tensors, padding=padding, truncation=truncation, max_length=max_length
+        )
```
Done

NielsRogge committed:
Remove vocab_size attribute

NielsRogge commented on src/transformers/models/llava_next/configuration_llava_next.py:
```diff
@@ -0,0 +1,147 @@
+            self.text_config = CONFIG_MAPPING[text_config["model_type"]](**text_config)
+            self.vocab_size = self.text_config.vocab_size
+        elif text_config is None:
+            self.text_config = CONFIG_MAPPING["llama"]()
```
Good point, I've removed `vocab_size` since this can be taken from `text_config`.

cc @ArthurZucker wondering why you added `vocab_size` as a separate attribute for llava

NielsRogge commented on src/transformers/models/llava_next/modeling_llava_next.py:
```diff
@@ -0,0 +1,684 @@
+        return self.language_model._supports_sdpa
+
+
+LLAVA_NEXT_INPUTS_DOCSTRING = r"""
```
Added those

NielsRogge committed:
Remove batch_size

NielsRogge committed:
Address comment

NielsRogge committed:
Add image processor tests

NielsRogge commented on docs/source/en/model_doc/llava_next.md:
```diff
@@ -0,0 +1,58 @@
+## Overview
+
+The LLaVA-NeXT model was proposed in [LLaVA-NeXT: Improved reasoning, OCR, and world knowledge](https://llava-vl.github.io/blog/2024-01-30-llava-next/) by Haotian Liu, Chunyuan Li, Yuheng Li, Bo Li, Yuanhan Zhang, Sheng Shen, Yong Jae Lee. LLaVa-NeXT (also called LLaVa-1.6) improves upon [LLaVa](llava) by increasing the input image resolution and training on an improved visual instruction tuning  ...
+
```
The model architecture is the same, the only differs is the increased image resolution. I've added a figure to the documentation for clarity

NielsRogge committed:
Support fx

NielsRogge commented on src/transformers/models/llava_next/image_processing_llava_next.py:
Tests are added

NielsRogge committed:
Update docstring

NielsRogge committed:
Add support for 34b

NielsRogge committed:
Convert 34b model

NielsRogge committed:
Add integration tests

NielsRogge committed:
Update checkpoints

NielsRogge commented:
Thanks for your review, addressed all comments and uploaded all checkpoints to a [collection](https://huggingface.co/collections/llava-hf/llava-next-65f75c4afac77fd37dbbe6cf).

The documentation test fails, seems like the machine crashed due to the size of the model (cc @ydshieh).

ydshieh commented:
> Thanks for your review, addressed all comments and uploaded all checkpoints to a [collection](https://huggingface.co/collections/llava-hf/llava-next-65f75c4afac77fd37dbbe6cf).
> 
> The documentation test fails, seems like the machine crashed due to the size of the model (cc @ydshieh).

If you can make use lower resource (anyway like fp16, 4-bit etc) that would be great.
If nothing could avoid the crash, just put the file to `utils/not_doctested.txt`

NielsRogge committed:
Convert vicuna-13b, remove doc tests

amyeroberts commented:
@NielsRogge Please review the PR before asking for review. There's files which should not be included as part of the PR here

NielsRogge committed:
Remove script

NielsRogge committed:
Remove file

NielsRogge commented:
Apologies, scripts are removed

amyeroberts commented on src/transformers/models/llava_next/configuration_llava_next.py:
```diff
@@ -0,0 +1,147 @@
+            The image token index to encode the image prompt.
+        projector_hidden_act (`str`, *optional*, defaults to `"gelu"`):
+            The activation function used by the multimodal projector.
+        vision_feature_select_strategy (`str`, *optional*, defaults to `"default"`):
```
OK - can you add an explanation here about what "default" versus "full" means?

NielsRogge replied:
Added

amyeroberts reviewed (commented):
Thanks for the continued work on this. It's almost ready, but some more coverage is needed for tests. In particular, the generation tests need to be added for the model

NielsRogge committed:
Address comments

NielsRogge committed:
Improve docstrings

NielsRogge commented on tests/models/llava_next/test_image_processor_llava_next.py:
Yes, we test both `aspect_ratio_setting = "clip"` and `aspect_ratio_setting = "anyres"`

NielsRogge commented on src/transformers/models/llava_next/image_processing_llava_next.py:
```diff
@@ -0,0 +1,704 @@
+
+            new_images.append(pixel_values)
+
+        if aspect_ratio_setting == "anyres":
+            data = {"pixel_values": new_images, "image_sizes": image_sizes}
+        elif aspect_ratio_setting in ["clip", "pad"]:
+            data = {"pixel_values": new_images}
```
That's a good point, actually `LlavaNextForConditionalGeneration` expects the `image_sizes` argument.

The other 2 settings ("clip" and "pad") are only there for people that want to have the correct image processing settings for `LlavaForConditionalGeneration`.

NielsRogge committed:
Deprecate vocab_size

NielsRogge commented on src/transformers/models/llava/configuration_llava.py:
```diff
@@ -116,15 +121,17 @@ def __init__(
-            self.text_config = CONFIG_MAPPING["llama"]()
+            text_config = CONFIG_MAPPING["llama"]()
+
+        self.vocab_size = vocab_size
```
Deprecated `vocab_size` as it can be taken from `text_config`

NielsRogge commented:
Regarding your comment regarding the `aspect_ratio_setting` of [LlavaNextImageProcessor](url) -> actually only `aspect_ratio_setting = "anyres"` is used for llava 1.6 (`LlavaNextForConditionalGeneration`).

The other 2 settings ("clip" and "pad") are there to be used for `LlavaForConditionalGeneration` (the previous generation). This one currently uses CLIPImageProcessor, which is technically incorrect, it should use the "pad" setting. But I assume we can't simply replace the image processor of `LlavaForConditionalGeneration` by a new one.

Should I keep the `aspect_ratio_setting`, or remove it and only support "anyres" by default?

amyeroberts commented:
> This one currently uses CLIPImageProcessor, which is technically incorrect, it should use the "pad" setting of LlavaNextImageProcessor. But I assume we can't simply replace the image processor of LlavaForConditionalGeneration by a new one (cc @ArthurZucker).

I don't quite understand this statement. "This one" is referring to `LlavaForConditionalGeneration` or `LlavaNextForConditionalGeneration`? 

> But I assume we can't simply replace the image processor of LlavaForConditionalGeneration by a new one (cc @ArthurZucker).

Again, not sure which object you're talking about replacing, but if there's an incorrect implementation it would be better to fix  it, possibly with some kind of deprecation cycle / legacy behaviour flag. If the issue is with the pad method, I don't see why we need to completely replace the image processor, instead of updating `pad`

> Should I keep the aspect_ratio_setting, or remove it and only support "anyres" by default?

We should remove the `aspect_ratio_setting` flag entirely

It doesn't make sense to have an image processor for a specific model, which is preparing inputs for a different model which already has its own image processor.

NielsRogge committed:
Remove aspect_ratio_setting

NielsRogge commented:
Ok, I've removed the flag as it leads to confusion. Feel free to approve the PR, if possible.

amyeroberts commented on src/transformers/models/llava_next/modeling_llava_next.py:
```diff
@@ -0,0 +1,690 @@
+        return self.language_model._supports_sdpa
+
+
+LLAVA_NEXT_INPUTS_DOCSTRING = r"""
```
docstring doesn't match signature

amyeroberts reviewed (commented):
Thanks for iterating! 

PR can't be merged as-is as it introduces breaking changes to existing models: llava and vip llava

NielsRogge commented on tests/models/llava/test_modeling_llava.py:
```diff
@@ -159,9 +159,7 @@ class LlavaForConditionalGenerationModelTest(ModelTesterMixin, unittest.TestCase
     pipeline_model_mapping = {"image-to-text": LlavaForConditionalGeneration} if is_torch_available() else {}
-    fx_compatible = False
     test_pruning = False
-    test_resize_embeddings = True
```
Because `test_resize_embeddings` is set to `True` by default in test_modeling_common.py

NielsRogge commented on tests/models/vipllava/test_modeling_vipllava.py:
```diff
@@ -185,6 +186,188 @@ def test_training_gradient_checkpointing_use_reentrant(self):
+            model_tied = model_class(config_tied)
+            params_tied = list(model_tied.parameters())
+            # Check that the embedding layer and decoding layer are the same in size and in value
+            # self.assertTrue(check_same_values(embeddings, decoding))
```
This is copied from the [common test](https://github.com/huggingface/transformers/blob/66ce9593fdb8e340df546ddd0774eb444f17a12c/tests/test_modeling_common.py#L1923)

NielsRogge commented on tests/models/llava/test_modeling_llava.py:
```diff
@@ -186,6 +184,188 @@ def test_training_gradient_checkpointing_use_reentrant(self):
     def test_training_gradient_checkpointing_use_reentrant_false(self):
         pass
 
+    # Copied from tests.test_modeling_common.ModelTesterMixin.test_resize_tokens_embeddings with config.vocab_size->config.text_config.vocab_size
```
I've tried that, but that doesn't work. The test fails with:
```python
# Check that resizing the token embeddings with a larger vocab size increases the model's vocab size
            model_embed = model.resize_token_embeddings(model_vocab_size + 10)
>           self.assertEqual(model.config.vocab_size, model_vocab_size + 10)
E           AssertionError: 99 != 109
```
This is because when calling the resize_token_embeddings method, we are changing `config.text_config.vocab_size`, whereas `config.vocab_size` remains unchanged.

amyeroberts replied:
That's because you removed the line that properly updates the config value in `resize_token_embeddings`

amyeroberts replied:
Yes, I agree we don't want to just add stuff to make the tests pass. I mistakenly though `vocab_size` was still being set in the config for BC, in which case, we would have still needed to update the parameter. 

If we leave as a property to deprecate (cf notes in vip llava config file) then we can leave copied from headers here

NielsRogge committed:
Address comments

NielsRogge committed:
Update READMEs

NielsRogge commented on tests/models/llava_next/test_modeling_llava_next.py:
```diff
@@ -0,0 +1,477 @@
+            model_tied = model_class(config_tied)
+            params_tied = list(model_tied.parameters())
+            # Check that the embedding layer and decoding layer are the same in size and in value
+            # self.assertTrue(check_same_values(embeddings, decoding))
```
Opened https://github.com/huggingface/transformers/pull/29737 to fix this

amyeroberts replied:
rebase and make fix-copies should fix this now :)

ArthurZucker commented on src/transformers/utils/dummy_pt_objects.py:
```diff
@@ -4796,7 +4796,17 @@ def __init__(self, *args, **kwargs):
         requires_backends(self, ["torch"])
 
 
-class LlavaProcessor(metaclass=DummyObject):
```
I don't thin we should fully remove the LlavaProcessor no?

NielsRogge replied:
I think this is because LlavaProcessor was at the wrong place in the main init. Multimodel processors don't require a backend like torch.

ArthurZucker commented on tests/models/musicgen_melody/test_modeling_musicgen_melody.py:
this is unrelated should be reverted

ArthurZucker reviewed (commented):
LGTM let's just fix the last 2 nits!

amyeroberts commented on src/transformers/models/llava/configuration_llava.py:
```diff
@@ -88,25 +88,36 @@ def __init__(
+                FutureWarning,
+            )
+            # set the vocab_size
+            text_config["vocab_size"] = kwargs["vocab_size"]
```
No - this is still wrong and will break backwards compatibility

amyeroberts replied:
vocab size from the kwargs was only ever set when text_config was `None`

NielsRogge replied:
Just pushed an update which is resolves this. If not, let me know how we can make it backwards compatible.

NielsRogge committed:
Add tips about chat templates

NielsRogge committed:
Fix tests

amyeroberts commented on src/transformers/models/vipllava/configuration_vipllava.py:
```diff
@@ -88,15 +87,21 @@ def __init__(
         self.vision_feature_layers = vision_feature_layers
-        self.vocab_size = vocab_size
+
+        if "vocab_size" in kwargs and not isinstance(text_config, dict):
+            warnings.warn(
+                "The `vocab_size` argument is deprecated and will be removed in v4.42, since it can be inferred from the `text_config`.",
+                FutureWarning,
+            )
+            # set the vocab_size
+            text_config["vocab_size"] = kwargs["vocab_size"]
```
Accidentally left? This has the same BC issues as before and conflicts with the code below

```suggestion
```

amyeroberts commented on src/transformers/models/llava/configuration_llava.py:
```diff
@@ -116,15 +119,22 @@ def __init__(
         elif text_config is None:
-            self.text_config = CONFIG_MAPPING["llama"]()
+            text_config = CONFIG_MAPPING["llama"]()
+            if "vocab_size" in kwargs:
+                warnings.warn(
+                    "The `vocab_size` argument is deprecated and will be removed in v4.42, since it can be inferred from the `text_config`.",
+                    FutureWarning,
+                )
+                # set the vocab_size
+                text_config.vocab_size = kwargs["vocab_size"]
```
Same comments as for vip llava:
* Just have a warning at the start of the init
* Add some deprecation for `vocab_size`

amyeroberts reviewed (commented):
Thanks for the continued iterations! 

Just a few final notes on BC for the config files

amyeroberts committed:
Deprecate vocab_size safely

amyeroberts committed:
Update tests

amyeroberts reviewed (approved):
Pushed the changes requested so all good to merge once tests pass!

ArthurZucker commented:
https://github.com/huggingface/transformers/commit/a2a95164b82960ee1b36b01735b81a8a48ee1958 there were some unprotected imports 😈

fxmarty commented on src/transformers/models/llava/configuration_llava.py:
```diff
@@ -88,25 +88,34 @@ def __init__(
+
         self.vision_feature_select_strategy = vision_feature_select_strategy
         self.vision_feature_layer = vision_feature_layer
-        self.vocab_size = vocab_size
```
```
self = <tests.models.llava.test_modeling_llava.LlavaForConditionalGenerationIntegrationTest testMethod=test_small_model_integration_test_llama>

    @slow
    @require_bitsandbytes
    def test_small_model_integration_test_llama(self):
        # Let' s make sure we test the preprocessing to replace what is used
        model_id = "llava-hf/llava-1.5-7b-hf"
    
>       model = LlavaForConditionalGeneration.from_pretrained("llava-hf/llava-1.5-7b-hf", load_in_4bit=True)

tests/models/llava/test_modeling_llava.py:390: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
src/transformers/modeling_utils.py:3006: in from_pretrained
    config, model_kwargs = cls.config_class.from_pretrained(
src/transformers/configuration_utils.py:609: in from_pretrained
    return cls.from_dict(config_dict, **kwargs)
src/transformers/configuration_utils.py:761: in from_dict
    config = cls(**config_dict)
src/transformers/models/llava/configuration_llava.py:140: in __init__
    super().__init__(**kwargs)
src/transformers/configuration_utils.py:375: in __init__
    raise err
src/transformers/configuration_utils.py:372: in __init__
    setattr(self, key, value)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = LlavaConfig {
  "architectures": [
    "LlavaForConditionalGeneration"
  ],
  "ignore_index": -100,
  "image_token_ind..._dim": 768,
    "vocab_size": 32000
  },
  "vision_feature_layer": -2,
  "vision_feature_select_strategy": "default"
}

key = 'vocab_size', value = 32064

    def __setattr__(self, key, value):
        if key in super().__getattribute__("attribute_map"):
            key = super().__getattribute__("attribute_map")[key]
>       super().__setattr__(key, value)
E       AttributeError: can't set attribute
```

fxmarty replied:
will fix in https://github.com/huggingface/transformers/pull/29389

amyeroberts replied:
sorry about this - thanks for addressing and fixing

aliencaocao commented:
This model when loading gives `The model weights are not tied. Please use the `tie_weights` method before using the `infer_auto_device` function.`. I am using device_map='auto'. Anyway to fix this?

amyeroberts commented:
Hi @aliencaocao, could you open an issue detailing this error and a minimal code snippet to reproduce? This will help us track what has and hasn't been addressed and help people find it if they're experiencing something similar.

aliencaocao commented:
https://github.com/huggingface/transformers/issues/30001
